sdk/nodejs/runtime/resource.ts

```diff
@@ -227,62 +233,139 @@ async function prepareResource(label: string, res: Resource, custom: boolean,
+    // the dependency graph and optimize operations accordingly.
+
+    // The list of all dependencies (implicit or explicit).
+    const allDirectDependencies = new Set<Resource>(explicitDirectDependencies);
+
+    const allDirectDependencyURNs = await getResourceURNs(explicitDirectDependencies);
+    const propertyToDirectDependencyURNs = new Map<string, Set<URN>>();
+
+    for (const [propertyName, directDependencies] of propertyToDirectDependencies) {
+        addAll(allDirectDependencies, directDependencies);
+
+        const urns = await getResourceURNs(directDependencies);
+        addAll(allDirectDependencyURNs, urns);
+        propertyToDirectDependencyURNs.set(propertyName, urns);
     }
 
+    // Now await all dependencies transitively.  i.e. if one resource depends on some other
+    // component resource, then it will end up depending on all the children of that component
+    // resource as well.  That way, a parent acts as a "logical" collection of all those children
+    // and will not be considered complete until all of the children are complete as well.
+    await getResourceURNs(getTransitivelyDependentResources(allDirectDependencies, res));
+
     return {
         resolveURN: resolveURN!,
         resolveID: resolveID,
         resolvers: resolvers,
         serializedProps: serializedProps,
         parentURN: parentURN,
         providerRef: providerRef,
-        dependencies: dependencies,
-        propertyDependencies: propertyDeps,
+        allDirectDependencyURNs: allDirectDependencyURNs,
+        propertyToDirectDependencyURNs: propertyToDirectDependencyURNs,
     };
 }
 
+function addAll<T>(to: Set<T>, from: Set<T>) {
+    for (const val of from) {
+        to.add(val);
+    }
+}
+
+async function getResourceURNs(resources: Set<Resource>) {
+    const result = new Set<URN>();
+    for (const resource of resources) {
+        result.add(await resource.urn.promise());
+    }
+
+    return result;
+}
+
+function getTransitivelyDependentResources(resources: Set<Resource>, thisResource: Resource) {
+    // keep iterating until we reach a fixed point.
+    while (true) {
+        const temp = new Set();
+
+        // Hit each resource, and then add all the resources it depends on.
+        for (const resource of resources) {
+            temp.add(resource);
+
+            if (resource.__childResources) {
+                for (const childResource of resource.__childResources) {
+                    resources.add(childResource);
+                }
+            }
+        }
+
+        // If the size didn't increase, then we found no new resources.  Can stop now.
+        if (temp.size === resources.size) {
+            break;
+        }
+
+        resources = temp;
+    }
+
+    // Now, exclude any resources that have 'this resource' somewhere in their parent chain.
+    // If we attempted to wait on them, we'd get a deadlock as they wait to wait on us.
+    const final = new Set<Resource>();
```

Comment:
Couldn't we achieve the same effect by skipping any resource that is equal to `thisResource` above?

Reply:
I don't believe so.  what if we simply just hit a resource that happens to be a descendent of us?  it can happen with all the async pumping that is going on.  recall that we compute this set *after* serializing the props-bag (which is done asynchronously).